File: applications/v3/frontend/src/app/core/models/api.models.ts
export type AccountRole = 'ADMIN' | 'VIEWER';
export type PersonStatus = 'ACTIVE' | 'INACTIVE';
export type ProjectStatus = 'ACTIVE' | 'INACTIVE';
export type WeekStatus = 'PLANNED' | 'COMPLETED';
export type DayOfWeek =
  | 'MONDAY'
  | 'TUESDAY'
  | 'WEDNESDAY'
  | 'THURSDAY'
  | 'FRIDAY'
  | 'SATURDAY'
  | 'SUNDAY';
export type SortDirection = 'asc' | 'desc';

export interface LoginRequest {
  username: string;
  password: string;
}

export interface LoginResponse {
  accessToken: string;
  tokenType: 'Bearer';
  expiresIn: number;
  account: AuthenticatedAccount;
}

export interface AuthenticatedAccount {
  accountId: string;
  personId: string;
  username: string;
  name: string;
  email: string;
  role: AccountRole;
  status: PersonStatus;
}

export interface AccountRegistrationRequest {
  name: string;
  username?: string;
  email: string;
  password: string;
}

export interface AccountRegistrationResponse {
  accountId: string;
  personId: string;
  username: string;
  name: string;
  email: string;
  role: AccountRole;
  status: PersonStatus;
  createdAt: string;
}

export interface PasswordChangeRequest {
  username: string;
  currentPassword: string;
  newPassword: string;
  confirmNewPassword: string;
}

export interface CreatePersonRequest {
  name: string;
  username?: string;
  email: string;
  password: string;
  role: AccountRole;
  status: PersonStatus;
}

export interface UpdatePersonRequest {
  name: string;
  username?: string;
  email: string;
  role: AccountRole;
  status: PersonStatus;
}

export interface PersonSummary {
  personId: string;
  accountId: string;
  username: string;
  name: string;
  email: string;
  role: AccountRole;
  status: PersonStatus;
}

export interface PersonDetail extends PersonSummary {
  createdAt: string;
  updatedAt: string;
}

export interface CreateProjectRequest {
  name: string;
  code: string;
  status: ProjectStatus;
}

export interface UpdateProjectRequest {
  name: string;
  code: string;
  status: ProjectStatus;
}

export interface ProjectSummary {
  projectId: string;
  name: string;
  code: string;
  status: ProjectStatus;
}

export interface ProjectDetail extends ProjectSummary {
  createdAt: string;
  updatedAt: string;
}

export interface CreateWeekRequest {
  weekStart: string;
  weekEnd: string;
  status?: WeekStatus;
}

export interface UpdateWeekRequest {
  weekStart: string;
  weekEnd: string;
  status: WeekStatus;
}

export interface WeekSummary {
  weekId: string;
  weekStart: string;
  weekEnd: string;
  status: WeekStatus;
}

export interface WeekDetail extends WeekSummary {
  createdAt: string;
  updatedAt: string;
}

export interface WeekPlanningBoard {
  week: WeekDetail;
  dayPlans: DayPlan[];
}

export interface DayPlanCollection {
  weekId: string;
  dayPlans: DayPlan[];
}

export interface DayPlan {
  dayPlanId: string;
  weekId: string;
  date: string;
  dayOfWeek: DayOfWeek;
  dayPersons: DayPerson[];
}

export interface DayPerson {
  dayPersonId: string;
  person: PersonReference;
  dayPersonProjects: DayPersonProject[];
}

export interface DayPersonProject {
  dayPersonProjectId: string;
  project: ProjectReference;
  tasks: Task[];
}

export interface Task {
  taskId: string;
  description: string;
  createdAt: string;
  updatedAt: string;
}

export interface PersonReference {
  personId: string;
  name: string;
  email: string;
  status: PersonStatus;
}

export interface ProjectReference {
  projectId: string;
  name: string;
  code: string;
  status: ProjectStatus;
}

export interface AddPersonToDayRequest {
  personId: string;
}

export interface AddProjectToDayPersonRequest {
  projectId: string;
}

export interface AddTaskRequest {
  description: string;
}

export interface SortOrder {
  property: string;
  direction: SortDirection;
}

export interface PageMetadata {
  page: number;
  size: number;
  totalElements: number;
  totalPages: number;
  sort?: SortOrder[];
}

export interface PersonPage {
  content: PersonSummary[];
  page: PageMetadata;
}

export interface ProjectPage {
  content: ProjectSummary[];
  page: PageMetadata;
}

export interface WeekPage {
  content: WeekSummary[];
  page: PageMetadata;
}

export interface FieldError {
  field: string;
  message: string;
  rejectedValue?: string | null;
}

export interface ErrorResponse {
  timestamp: string;
  status: number;
  error: string;
  code: string;
  message: string;
  path: string;
  fieldErrors?: FieldError[];
}

File: applications/v2/backend/src/main/java/com/weeklyplanning/web/rest/api/ApiModels.java
package com.weeklyplanning.web.rest.api;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.weeklyplanning.domain.enumeration.AccountRole;
import com.weeklyplanning.domain.enumeration.PersonStatus;
import com.weeklyplanning.domain.enumeration.ProjectStatus;
import com.weeklyplanning.domain.enumeration.WeekStatus;
import jakarta.validation.constraints.Email;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Size;
import java.time.LocalDate;
import java.time.OffsetDateTime;
import java.util.List;
import java.util.UUID;

public final class ApiModels {

    private ApiModels() {}

    public record LoginRequest(@NotBlank @Size(min = 3, max = 50) String username, @NotBlank @Size(min = 8, max = 128) String password) {}

    public record AccountCreateRequest(
        @NotBlank @Size(min = 3, max = 50) String username,
        @NotBlank @Size(min = 8, max = 128) String password,
        @NotNull AccountRole role
    ) {}

    public record PersonUpsertRequest(
        @NotBlank @Size(min = 1, max = 100) String name,
        @NotBlank @Email @Size(max = 254) String email,
        @NotBlank @Size(min = 1, max = 50) String role,
        @NotNull PersonStatus status
    ) {}

    public record ProjectUpsertRequest(
        @NotBlank @Size(min = 1, max = 120) String name,
        @NotBlank @Size(min = 1, max = 20) String code,
        @NotNull UUID ownerId,
        @NotNull ProjectStatus status
    ) {}

    public record WeekUpsertRequest(@NotNull LocalDate startDate, @NotNull LocalDate endDate, @NotNull WeekStatus status) {}

    public record AddPersonToDayRequest(@NotNull UUID personId) {}

    public record AddProjectToPersonRequest(@NotNull UUID projectId) {}

    public record AddTaskRequest(@NotBlank @Size(min = 1, max = 1000) String description) {}

    public record AccountSummary(UUID id, String username, AccountRole role) {}

    public record LoginResponse(String accessToken, String tokenType, @JsonInclude(JsonInclude.Include.NON_NULL) Long expiresIn, AccountSummary account) {}

    public record AccountResponse(UUID id, String username, AccountRole role) {}

    public record Person(UUID id, String name, String email, String role, PersonStatus status) {}

    public record PersonSummary(UUID id, String name) {}

    public record Project(UUID id, String name, String code, PersonSummary owner, ProjectStatus status) {}

    public record ProjectSummary(UUID id, String name, String code) {}

    public record Week(UUID id, LocalDate startDate, LocalDate endDate, WeekStatus status) {}

    public record Task(UUID id, String description) {}

    public record DayPersonProject(ProjectSummary project, List<Task> tasks) {}

    public record DayPerson(PersonSummary person, List<DayPersonProject> projects) {}

    public record DayPlan(LocalDate date, List<DayPerson> people) {}

    public record WeekPlanning(UUID id, LocalDate startDate, LocalDate endDate, WeekStatus status, List<DayPlan> days) {}

    public enum SortDirection {
        ASC,
        DESC,
    }

    public record SortInfo(String property, SortDirection direction) {}

    public record PageInfo(int page, int size, long totalElements, int totalPages, @JsonInclude(JsonInclude.Include.NON_NULL) List<SortInfo> sort) {}

    public record PeoplePage(List<Person> items, PageInfo page) {}

    public record ProjectsPage(List<Project> items, PageInfo page) {}

    public record WeeksPage(List<Week> items, PageInfo page) {}

    public record FieldError(String field, String message) {}

    public record ErrorResponse(
        OffsetDateTime timestamp,
        int status,
        String error,
        String message,
        String path,
        @JsonInclude(JsonInclude.Include.NON_EMPTY) List<FieldError> details
    ) {}
}
